projects/extension/src/background/index.ts

```diff
@@ -24,9 +24,46 @@ export interface Background extends Window {
       listener: (state: ExposedChainConnection[]) => void,
     ) => () => void
     disconnectTab: (tabId: number) => void
+    getLogger: () => LogKeeper
   }
 }
 
+interface LogKeeper {
+  debug: string[]
+  warn: string[]
+  error: string[]
+}
+
+const logKeeper: LogKeeper = {
+  debug: [],
+  warn: [],
+  error: [],
+}
+
+const getTime = () => {
+  const date = new Date()
+  return `${date.getHours() < 10 ? "0" : ""}${date.getHours()}:${
+    date.getMinutes() < 10 ? "0" : ""
+  }${date.getMinutes()}:${
+    date.getSeconds() < 10 ? "0" : ""
+  }${date.getSeconds()} ${date.getMilliseconds()}`
+}
+
+const logger = (level: number, target: string, message: string) => {
+  const logLevels = ["Error", "Warn", "Info", "Debug"]
+  if (level !== 4) {
+    // log all non-debug logs to background console
+    console.log(`${logLevels[level - 1]}: ${message}`)
+    if (level === 1) {
+      logKeeper.error.push(`${getTime()} | ${target} | ${message}`)
+    } else if (level === 2) {
+      logKeeper.warn.push(`${getTime()} | ${target} | ${message}`)
+    }
+  }
+  logKeeper.debug.length >= 1000 && logKeeper.debug.shift()
```

Comment:
```suggestion
  if (logKeeper.debug.length >= 1000) logKeeper.debug.shift()
```

Reply:
Also, I think it's a good idea to do the same for warnings and errors, just in case